File: src/database/queries.ts
import { DatabaseConnection } from './connection.js';
import { QueryResult, TableInfo, TableSchema, ColumnInfo, QueryOptions } from '../types/index.js';

interface SecurityValidationResult {
  isValid: boolean;
  error?: string;
}

export class DatabaseQueries {
  private db: DatabaseConnection;

  // List of allowed SQL keywords for read-only operations
  private readonly ALLOWED_KEYWORDS = [
    'SELECT', 'WITH', 'FROM', 'WHERE', 'JOIN', 'INNER', 'LEFT', 'RIGHT', 'FULL', 'OUTER',
    'GROUP', 'BY', 'HAVING', 'ORDER', 'LIMIT', 'OFFSET', 'UNION', 'INTERSECT', 'EXCEPT',
    'AS', 'AND', 'OR', 'NOT', 'IN', 'EXISTS', 'BETWEEN', 'LIKE', 'ILIKE', 'SIMILAR',
    'CASE', 'WHEN', 'THEN', 'ELSE', 'END', 'CAST', 'EXTRACT', 'COUNT', 'SUM', 'AVG',
    'MIN', 'MAX', 'DISTINCT', 'ALL', 'ANY', 'SOME', 'TRUE', 'FALSE', 'NULL', 'IS'
  ];

  // List of forbidden SQL statement starters (mutating operations)
  private readonly FORBIDDEN_STATEMENTS = [
    'INSERT', 'UPDATE', 'DELETE', 'DROP', 'CREATE', 'ALTER', 'TRUNCATE', 'REPLACE',
    'MERGE', 'UPSERT', 'COPY', 'BULK', 'GRANT', 'REVOKE', 'COMMIT', 'ROLLBACK',
    'BEGIN', 'START', 'SAVEPOINT', 'RELEASE', 'SET', 'RESET',
    'SHOW', 'EXPLAIN', 'ANALYZE', 'VACUUM', 'REINDEX', 'LOCK', 'UNLOCK'
  ];

  // List of forbidden SQL commands that could appear anywhere
  private readonly FORBIDDEN_COMMANDS = [
    'CLUSTER INDEX', 'CLUSTER TABLE', 'DROP TABLE', 'DROP INDEX', 'CREATE TABLE',
    'CREATE INDEX', 'ALTER TABLE', 'ALTER INDEX', 'TRUNCATE TABLE'
  ];

  constructor(db: DatabaseConnection) {
    this.db = db;
  }

  /**
   * Validates SQL query for security and read-only compliance
   */
  private validateQuery(sql: string): SecurityValidationResult {
    // Normalize the SQL query
    const normalizedSql = sql.trim().toUpperCase();

    // Check if query is empty
    if (!normalizedSql) {
      return { isValid: false, error: 'Empty query not allowed' };
    }

    // Check if query starts with forbidden statement types
    for (const statement of this.FORBIDDEN_STATEMENTS) {
      if (normalizedSql.startsWith(statement + ' ') || normalizedSql === statement) {
        return {
          isValid: false,
          error: `Mutating operation '${statement}' is not allowed. This server is read-only.`
        };
      }
    }

    // Check for forbidden multi-word commands anywhere in the query
    for (const command of this.FORBIDDEN_COMMANDS) {
      if (normalizedSql.includes(command)) {
        return {
          isValid: false,
          error: `Operation '${command}' is not allowed. This server is read-only.`
        };
      }
    }

    // Check for SQL injection patterns
    const sqlInjectionPatterns = [
      /;[\s]*(--)/, // SQL comments after semicolon
      /;[\s]*(\/\*)/, // Block comments after semicolon
      /[\s]+(OR|AND)[\s]+1[\s]*=[\s]*1/i, // Classic OR 1=1 injection
      /[\s]+(OR|AND)[\s]+\w+[\s]*=[\s]*\w+/i, // Field=field injection
      /UNION[\s]+ALL[\s]+SELECT/i, // UNION injection
      /[\s]+UNION[\s]+SELECT/i, // UNION injection
      /'[\s]*(OR|AND)[\s]*'/i, // Quote-based injection
      /--[\s]*$/, // SQL comments at end
      /\/\*.*\*\//s, // Block comments
      /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]/, // Control characters
      /exec[\s]*\(/i, // EXEC function calls
      /sp_/i, // Stored procedure calls
      /xp_/i, // Extended stored procedures
    ];

    for (const pattern of sqlInjectionPatterns) {
      if (pattern.test(sql)) {
        return {
          isValid: false,
          error: 'Query contains potentially unsafe patterns and has been blocked for security.'
        };
      }
    }

    // Ensure query starts with SELECT or WITH (for CTEs)
    if (!normalizedSql.startsWith('SELECT') && !normalizedSql.startsWith('WITH')) {
      return {
        isValid: false,
        error: 'Only SELECT queries and CTEs (WITH) are allowed.'
      };
    }

    // Additional validation: Check for multiple statements
    const statements = sql.split(';').filter(s => s.trim().length > 0);
    if (statements.length > 1) {
      return {
        isValid: false,
        error: 'Multiple SQL statements are not allowed. Please execute one query at a time.'
      };
    }

    return { isValid: true };
  }

  async executeQuery(sql: string, parameters?: any[], options?: QueryOptions): Promise<QueryResult> {
    try {
      // Validate query for security and read-only compliance
      const validation = this.validateQuery(sql);
      if (!validation.isValid) {
        throw new Error(`Security validation failed: ${validation.error}`);
      }

      // only log when debugging is enabled
      if (process.env.DEBUG === 'true') {
        console.log('[QUERY DEBUG] SQL:', sql);
        console.log('[QUERY DEBUG] Parameters:', parameters);
      }

      const result = await this.db.query(sql, parameters);
      
      // log result structure for debugging
      if (process.env.DEBUG === 'true') {
        console.log('[QUERY DEBUG] Query result structure:', {
          hasRows: !!result.rows,
          rowsType: typeof result.rows,
          rowsIsArray: Array.isArray(result.rows),
          rowsLength: result.rows?.length,
          fields: result.fields?.length,
          rowCount: result.rowCount
        });
      }
      
      // Convert row objects to arrays if needed
      const rowsAsArrays = result.rows.map(row => {
        if (Array.isArray(row)) {
          return row;
        } else {
          // Convert object to array based on field order
          return result.fields.map(field => row[field.name]);
        }
      });
      
      const queryResult: QueryResult = {
        columns: result.fields.map(field => field.name),
        rows: rowsAsArrays,
        rowCount: result.rowCount || 0,
        executionTime: (result as any).executionTime
      };

      // Apply row limit if specified
      if (options?.maxRows && queryResult.rows.length > options.maxRows) {
        queryResult.rows = queryResult.rows.slice(0, options.maxRows);
      }

      return queryResult;
    } catch (error) {
      console.error('Query execution failed:', error);
      throw new Error(`Query failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  async listTables(): Promise<TableInfo[]> {
    const sql = `
      SELECT 
        schemaname as schema,
        tablename as table_name
      FROM pg_tables 
      WHERE schemaname IN ('public', 'search')
      ORDER BY schemaname, tablename
    `;

    const result = await this.executeQuery(sql);
    
    return result.rows.map(row => ({
      tableName: row[1], // tablename
      schema: row[0],    // schemaname
      rowCount: undefined // We'll get this separately if needed
    }));
  }

  async describeTable(tableName: string, schema: string = 'public'): Promise<TableSchema> {
    // Get column information
    const columnSql = `
      SELECT 
        column_name,
        data_type,
        is_nullable,
        column_default,
        col_description((table_schema||'.'||table_name)::regclass, ordinal_position) as description
      FROM information_schema.columns 
      WHERE table_schema = $1 AND table_name = $2
      ORDER BY ordinal_position
    `;

    const columnResult = await this.executeQuery(columnSql, [schema, tableName]);
    
    const columns: ColumnInfo[] = columnResult.rows.map(row => ({
      columnName: row[0],
      dataType: row[1],
      isNullable: row[2] === 'YES',
      defaultValue: row[3] || undefined,
      description: row[4] || undefined
    }));

    // Get index information
    const indexSql = `
      SELECT indexname 
      FROM pg_indexes 
      WHERE schemaname = $1 AND tablename = $2
    `;

    const indexResult = await this.executeQuery(indexSql, [schema, tableName]);
    const indexes = indexResult.rows.map(row => row[0]);

    // Get constraint information
    const constraintSql = `
      SELECT 
        tc.constraint_name,
        tc.constraint_type,
        kcu.column_name
      FROM information_schema.table_constraints tc
      JOIN information_schema.key_column_usage kcu 
        ON tc.constraint_name = kcu.constraint_name
      WHERE tc.table_schema = $1 AND tc.table_name = $2
    `;

    const constraintResult = await this.executeQuery(constraintSql, [schema, tableName]);
    const constraints = constraintResult.rows.map(row => 
      `${row[1]} ${row[0]} (${row[2]})`
    );

    return {
      tableName,
      schema,
      columns,
      indexes,
      constraints
    };
  }

  async getTableData(tableName: string, schema: string = 'public', limit: number = 10): Promise<QueryResult> {
    const sql = `SELECT * FROM ${schema}.${tableName} LIMIT $1`;
    return await this.executeQuery(sql, [limit]);
  }

  async getTableRowCount(tableName: string, schema: string = 'public'): Promise<number> {
    const sql = `SELECT COUNT(*) as count FROM ${schema}.${tableName}`;
    const result = await this.executeQuery(sql);
    return parseInt(result.rows[0][0]);
  }

  async getTableSize(tableName: string, schema: string = 'public'): Promise<string> {
    const sql = `
      SELECT pg_size_pretty(pg_total_relation_size($1::regclass)) as size
    `;
    const result = await this.executeQuery(sql, [`${schema}.${tableName}`]);
    return result.rows[0][0];
  }

  async searchTables(searchTerm: string): Promise<TableInfo[]> {
    const sql = `
      SELECT 
        schemaname as schema,
        tablename as table_name
      FROM pg_tables 
      WHERE tablename ILIKE $1 OR schemaname ILIKE $1
      ORDER BY schemaname, tablename
    `;
    
    const result = await this.executeQuery(sql, [`%${searchTerm}%`]);
    
    return result.rows.map(row => ({
      tableName: row[1],
      schema: row[0]
    }));
  }

  async getDatabaseStats(): Promise<{
    tableCount: number;
    totalRows: number;
    databaseSize: string;
    searchSchemaSize: string;
    resourcesTableSize: string;
    edgesTableSize: string;
    activeConnections: number;
  }> {
    // Get table count - include both public and search schemas
    const tableCountSql = `
      SELECT COUNT(*) as count
      FROM pg_tables
      WHERE schemaname IN ('public', 'search')
    `;
    const tableCountResult = await this.executeQuery(tableCountSql);
    const tableCount = parseInt(tableCountResult.rows[0][0]);

    // Get total rows from ACM search tables specifically
    let totalRows = 0;
    try {
      const totalRowsSql = `
        SELECT
          (SELECT COUNT(*) FROM search.resources) +
          (SELECT COUNT(*) FROM search.edges) as total_rows
      `;
      const totalRowsResult = await this.executeQuery(totalRowsSql);
      totalRows = parseInt(totalRowsResult.rows[0][0]) || 0;
    } catch (error) {
      // Fallback to pg_stat_user_tables if search tables don't exist
      try {
        const fallbackSql = `
          SELECT COALESCE(SUM(n_live_tup), 0) as total_rows
          FROM pg_stat_user_tables
          WHERE schemaname IN ('public', 'search')
        `;
        const fallbackResult = await this.executeQuery(fallbackSql);
        totalRows = parseInt(fallbackResult.rows[0][0]) || 0;
      } catch (fallbackError) {
        totalRows = 0;
      }
    }

    // Get database size
    const sizeSql = `SELECT pg_size_pretty(pg_database_size(current_database())) as size`;
    const sizeResult = await this.executeQuery(sizeSql);
    const databaseSize = sizeResult.rows[0][0];

    // Get search schema specific sizes
    let searchSchemaSize = 'N/A';
    let resourcesTableSize = 'N/A';
    let edgesTableSize = 'N/A';

    try {
      // Get total size of all tables in search schema
      const searchSchemaSql = `
        SELECT pg_size_pretty(
          COALESCE(SUM(pg_total_relation_size(schemaname||'.'||tablename)), 0)
        ) as size
        FROM pg_tables
        WHERE schemaname = 'search'
      `;
      const searchSchemaResult = await this.executeQuery(searchSchemaSql);
      searchSchemaSize = searchSchemaResult.rows[0][0];

      // Get individual table sizes
      const resourcesSizeSql = `
        SELECT pg_size_pretty(pg_total_relation_size('search.resources')) as size
      `;
      const resourcesSizeResult = await this.executeQuery(resourcesSizeSql);
      resourcesTableSize = resourcesSizeResult.rows[0][0];

      const edgesSizeSql = `
        SELECT pg_size_pretty(pg_total_relation_size('search.edges')) as size
      `;
      const edgesSizeResult = await this.executeQuery(edgesSizeSql);
      edgesTableSize = edgesSizeResult.rows[0][0];
    } catch (error) {
      // If search tables don't exist, sizes will remain 'N/A'
      console.error('Error getting search schema sizes:', error);
    }

    // Get active connections
    const connectionsSql = `
      SELECT COUNT(*) as count
      FROM pg_stat_activity
      WHERE state = 'active'
    `;
    const connectionsResult = await this.executeQuery(connectionsSql);
    const activeConnections = parseInt(connectionsResult.rows[0][0]);

    return {
      tableCount,
      totalRows,
      databaseSize,
      searchSchemaSize,
      resourcesTableSize,
      edgesTableSize,
      activeConnections
    };
  }
} 

File: golang/pkg/database/queries.go
// Package database provides database query functionality with read-only validation
package database

import (
	"context"
	"errors"
	"fmt"
	"log"
	"strconv"
	"strings"
	"time"

	"github.com/stolostron/search-mcp-server/pkg/types"
)

// SecurityValidationResult represents the result of SQL security validation
type SecurityValidationResult struct {
	IsValid bool
	Error   string
}

// DatabaseQueries provides database query functionality with security validation
type DatabaseQueries struct {
	db *DatabaseConnection

	// List of allowed SQL keywords for read-only operations
	allowedKeywords []string

	// List of forbidden SQL statement starters (mutating operations)
	forbiddenStatements []string

	// List of forbidden SQL commands that could appear anywhere
	forbiddenCommands []string
}

// NewDatabaseQueries creates a new DatabaseQueries instance
func NewDatabaseQueries(db *DatabaseConnection) *DatabaseQueries {
	dq := &DatabaseQueries{
		db: db,
		allowedKeywords: []string{
			"SELECT", "WITH", "FROM", "WHERE", "JOIN", "INNER", "LEFT", "RIGHT", "FULL", "OUTER",
			"GROUP", "BY", "HAVING", "ORDER", "LIMIT", "OFFSET", "UNION", "INTERSECT", "EXCEPT",
			"AS", "AND", "OR", "NOT", "IN", "EXISTS", "BETWEEN", "LIKE", "ILIKE", "SIMILAR",
			"CASE", "WHEN", "THEN", "ELSE", "END", "CAST", "EXTRACT", "COUNT", "SUM", "AVG",
			"MIN", "MAX", "DISTINCT", "ALL", "ANY", "SOME", "TRUE", "FALSE", "NULL", "IS",
		},
		forbiddenStatements: []string{
			"INSERT", "UPDATE", "DELETE", "DROP", "CREATE", "ALTER", "TRUNCATE", "REPLACE",
			"MERGE", "UPSERT", "COPY", "BULK", "GRANT", "REVOKE", "COMMIT", "ROLLBACK",
			"BEGIN", "START", "SAVEPOINT", "RELEASE", "SET", "RESET",
			"SHOW", "EXPLAIN", "ANALYZE", "VACUUM", "REINDEX", "LOCK", "UNLOCK",
		},
		forbiddenCommands: []string{
			"CLUSTER INDEX", "CLUSTER TABLE", "DROP TABLE", "DROP INDEX", "CREATE TABLE",
			"CREATE INDEX", "ALTER TABLE", "ALTER INDEX", "TRUNCATE TABLE",
		},
	}


	return dq
}

// validateQuery validates SQL query for read-only compliance
// NOTE: SQL injection protection comes from parameterized queries, NOT pattern detection
func (dq *DatabaseQueries) validateQuery(sql string) SecurityValidationResult {
	// Normalize the SQL query
	normalizedSQL := strings.TrimSpace(strings.ToUpper(sql))

	// Check if query is empty
	if normalizedSQL == "" {
		return SecurityValidationResult{
			IsValid: false,
			Error:   "Empty query not allowed",
		}
	}

	// Check if query starts with forbidden statement types
	for _, statement := range dq.forbiddenStatements {
		if strings.HasPrefix(normalizedSQL, statement+" ") || normalizedSQL == statement {
			return SecurityValidationResult{
				IsValid: false,
				Error:   fmt.Sprintf("Mutating operation '%s' is not allowed. This server is read-only.", statement),
			}
		}
	}

	// Check for forbidden multi-word commands anywhere in the query
	for _, command := range dq.forbiddenCommands {
		if strings.Contains(normalizedSQL, command) {
			return SecurityValidationResult{
				IsValid: false,
				Error:   fmt.Sprintf("Operation '%s' is not allowed. This server is read-only.", command),
			}
		}
	}


	// Ensure query starts with SELECT or WITH (for CTEs)
	if !strings.HasPrefix(normalizedSQL, "SELECT") && !strings.HasPrefix(normalizedSQL, "WITH") {
		return SecurityValidationResult{
			IsValid: false,
			Error:   "Only SELECT queries and CTEs (WITH) are allowed.",
		}
	}

	// Additional validation: Check for multiple statements
	statements := strings.Split(sql, ";")
	nonEmptyStatements := 0
	for _, stmt := range statements {
		if strings.TrimSpace(stmt) != "" {
			nonEmptyStatements++
		}
	}
	if nonEmptyStatements > 1 {
		return SecurityValidationResult{
			IsValid: false,
			Error:   "Multiple SQL statements are not allowed. Please execute one query at a time.",
		}
	}

	return SecurityValidationResult{IsValid: true}
}

// ExecuteQuery executes a SQL query with security validation and optional parameters
func (dq *DatabaseQueries) ExecuteQuery(ctx context.Context, sql string, parameters []interface{}, options *types.QueryOptions) (*types.QueryResult, error) {
	// Validate query for security and read-only compliance
	validation := dq.validateQuery(sql)
	if !validation.IsValid {
		return nil, fmt.Errorf("security validation failed: %s", validation.Error)
	}

	// Apply timeout if specified in options
	queryCtx := ctx
	var cancel context.CancelFunc
	if options != nil && options.Timeout != nil {
		queryCtx, cancel = context.WithTimeout(ctx, time.Duration(*options.Timeout)*time.Second)
		defer cancel()
	}

	result, err := dq.db.Query(queryCtx, sql, parameters...)
	if err != nil {
		// Check for timeout error and provide a clear message
		if errors.Is(err, context.DeadlineExceeded) {
			return nil, fmt.Errorf("query timeout exceeded (%d seconds)", *options.Timeout)
		}
		log.Printf("Query execution failed: %v", err)
		return nil, fmt.Errorf("query failed: %w", err)
	}

	// Apply row limit if specified
	if options != nil && options.MaxRows != nil && len(result.Rows) > *options.MaxRows {
		result.Rows = result.Rows[:*options.MaxRows]
		// Update row count to reflect the limit
		limitedCount := *options.MaxRows
		result.RowCount = &limitedCount
	}

	return result, nil
}

// ListTables returns a list of tables in public and search schemas
func (dq *DatabaseQueries) ListTables(ctx context.Context) ([]types.TableInfo, error) {
	sql := `
		SELECT
			schemaname as schema,
			tablename as table_name
		FROM pg_tables
		WHERE schemaname IN ('public', 'search')
		ORDER BY schemaname, tablename
	`

	result, err := dq.ExecuteQuery(ctx, sql, nil, nil)
	if err != nil {
		return nil, err
	}

	var tables []types.TableInfo
	for _, row := range result.Rows {
		if len(row) >= 2 {
			schema, schemaOk := row[0].(string)
			tableName, tableOk := row[1].(string)
			if schemaOk && tableOk {
				tables = append(tables, types.TableInfo{
					Schema:    schema,
					TableName: tableName,
				})
			}
		}
	}

	return tables, nil
}

// DescribeTable returns the schema details of a specified table
func (dq *DatabaseQueries) DescribeTable(ctx context.Context, tableName, schema string) (*types.TableSchema, error) {
	if schema == "" {
		schema = "public"
	}

	// Get column information
	columnSQL := `
		SELECT
			column_name,
			data_type,
			is_nullable,
			column_default,
			col_description((table_schema||'.'||table_name)::regclass, ordinal_position) as description
		FROM information_schema.columns
		WHERE table_schema = $1 AND table_name = $2
		ORDER BY ordinal_position
	`

	columnResult, err := dq.ExecuteQuery(ctx, columnSQL, []interface{}{schema, tableName}, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get column information: %w", err)
	}

	var columns []types.ColumnInfo
	for _, row := range columnResult.Rows {
		if len(row) >= 5 {
			columnName, columnOk := row[0].(string)
			dataType, typeOk := row[1].(string)
			isNullableStr, nullableOk := row[2].(string)

			if columnOk && typeOk && nullableOk {
				column := types.ColumnInfo{
					ColumnName: columnName,
					DataType:   dataType,
					IsNullable: isNullableStr == "YES",
				}

				// Handle nullable fields
				if row[3] != nil {
					if defaultVal, ok := row[3].(string); ok {
						column.DefaultValue = &defaultVal
					}
				}
				if row[4] != nil {
					if desc, ok := row[4].(string); ok {
						column.Description = &desc
					}
				}

				columns = append(columns, column)
			}
		}
	}

	// Get index information
	indexSQL := `
		SELECT indexname
		FROM pg_indexes
		WHERE schemaname = $1 AND tablename = $2
	`

	indexResult, err := dq.ExecuteQuery(ctx, indexSQL, []interface{}{schema, tableName}, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get index information: %w", err)
	}

	var indexes []string
	for _, row := range indexResult.Rows {
		if len(row) >= 1 {
			if indexName, ok := row[0].(string); ok {
				indexes = append(indexes, indexName)
			}
		}
	}

	// Get constraint information
	constraintSQL := `
		SELECT
			tc.constraint_name,
			tc.constraint_type,
			kcu.column_name
		FROM information_schema.table_constraints tc
		JOIN information_schema.key_column_usage kcu
			ON tc.constraint_name = kcu.constraint_name
		WHERE tc.table_schema = $1 AND tc.table_name = $2
	`

	constraintResult, err := dq.ExecuteQuery(ctx, constraintSQL, []interface{}{schema, tableName}, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get constraint information: %w", err)
	}

	var constraints []string
	for _, row := range constraintResult.Rows {
		if len(row) >= 3 {
			constraintName, nameOk := row[0].(string)
			constraintType, typeOk := row[1].(string)
			columnName, columnOk := row[2].(string)
			if nameOk && typeOk && columnOk {
				constraint := fmt.Sprintf("%s %s (%s)", constraintType, constraintName, columnName)
				constraints = append(constraints, constraint)
			}
		}
	}

	return &types.TableSchema{
		TableName:   tableName,
		Schema:      schema,
		Columns:     columns,
		Indexes:     indexes,
		Constraints: constraints,
	}, nil
}

// GetTableData returns sample data from a specified table
func (dq *DatabaseQueries) GetTableData(ctx context.Context, tableName, schema string, limit int) (*types.QueryResult, error) {
	if schema == "" {
		schema = "public"
	}

	sql := fmt.Sprintf("SELECT * FROM %s.%s LIMIT $1", schema, tableName)
	return dq.ExecuteQuery(ctx, sql, []interface{}{limit}, nil)
}

// GetTableRowCount returns the number of rows in a specified table
func (dq *DatabaseQueries) GetTableRowCount(ctx context.Context, tableName, schema string) (int, error) {
	if schema == "" {
		schema = "public"
	}

	sql := fmt.Sprintf("SELECT COUNT(*) as count FROM %s.%s", schema, tableName)
	result, err := dq.ExecuteQuery(ctx, sql, nil, nil)
	if err != nil {
		return 0, err
	}

	if len(result.Rows) > 0 && len(result.Rows[0]) > 0 {
		if count, ok := result.Rows[0][0].(int64); ok {
			return int(count), nil
		}
		// Try string conversion as fallback
		if countStr, ok := result.Rows[0][0].(string); ok {
			if count, err := strconv.Atoi(countStr); err == nil {
				return count, nil
			}
		}
	}

	return 0, fmt.Errorf("failed to get row count")
}

// GetTableSize returns the disk size of a specified table
func (dq *DatabaseQueries) GetTableSize(ctx context.Context, tableName, schema string) (string, error) {
	if schema == "" {
		schema = "public"
	}

	sql := `SELECT pg_size_pretty(pg_total_relation_size($1::regclass)) as size`
	fullTableName := fmt.Sprintf("%s.%s", schema, tableName)

	result, err := dq.ExecuteQuery(ctx, sql, []interface{}{fullTableName}, nil)
	if err != nil {
		return "", err
	}

	if len(result.Rows) > 0 && len(result.Rows[0]) > 0 {
		if size, ok := result.Rows[0][0].(string); ok {
			return size, nil
		}
	}

	return "", fmt.Errorf("failed to get table size")
}

// SearchTables searches for tables by name pattern
func (dq *DatabaseQueries) SearchTables(ctx context.Context, searchTerm string) ([]types.TableInfo, error) {
	sql := `
		SELECT
			schemaname as schema,
			tablename as table_name
		FROM pg_tables
		WHERE tablename ILIKE $1 OR schemaname ILIKE $1
		ORDER BY schemaname, tablename
	`

	searchPattern := fmt.Sprintf("%%%s%%", searchTerm)
	result, err := dq.ExecuteQuery(ctx, sql, []interface{}{searchPattern}, nil)
	if err != nil {
		return nil, err
	}

	var tables []types.TableInfo
	for _, row := range result.Rows {
		if len(row) >= 2 {
			schema, schemaOk := row[0].(string)
			tableName, tableOk := row[1].(string)
			if schemaOk && tableOk {
				tables = append(tables, types.TableInfo{
					Schema:    schema,
					TableName: tableName,
				})
			}
		}
	}

	return tables, nil
}

// DatabaseStatsResult represents the result of database statistics query
type DatabaseStatsResult struct {
	TableCount           int    `json:"tableCount"`
	TotalRows           int    `json:"totalRows"`
	DatabaseSize        string `json:"databaseSize"`
	SearchSchemaSize    string `json:"searchSchemaSize"`
	ResourcesTableSize  string `json:"resourcesTableSize"`
	EdgesTableSize      string `json:"edgesTableSize"`
	ActiveConnections   int    `json:"activeConnections"`
}

// GetDatabaseStats returns comprehensive database statistics
func (dq *DatabaseQueries) GetDatabaseStats(ctx context.Context) (*DatabaseStatsResult, error) {
	// Get table count - include both public and search schemas
	tableCountSQL := `
		SELECT COUNT(*) as count
		FROM pg_tables
		WHERE schemaname IN ('public', 'search')
	`
	tableCountResult, err := dq.ExecuteQuery(ctx, tableCountSQL, nil, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get table count: %w", err)
	}

	tableCount := 0
	if len(tableCountResult.Rows) > 0 && len(tableCountResult.Rows[0]) > 0 {
		if count, ok := tableCountResult.Rows[0][0].(int64); ok {
			tableCount = int(count)
		}
	}

	// Get total rows from ACM search tables specifically
	totalRows := 0
	totalRowsSQL := `
		SELECT
			(SELECT COUNT(*) FROM search.resources) +
			(SELECT COUNT(*) FROM search.edges) as total_rows
	`
	if totalRowsResult, err := dq.ExecuteQuery(ctx, totalRowsSQL, nil, nil); err == nil {
		if len(totalRowsResult.Rows) > 0 && len(totalRowsResult.Rows[0]) > 0 {
			if count, ok := totalRowsResult.Rows[0][0].(int64); ok {
				totalRows = int(count)
			}
		}
	} else {
		// Fallback to pg_stat_user_tables if search tables don't exist
		fallbackSQL := `
			SELECT COALESCE(SUM(n_live_tup), 0) as total_rows
			FROM pg_stat_user_tables
			WHERE schemaname IN ('public', 'search')
		`
		if fallbackResult, err := dq.ExecuteQuery(ctx, fallbackSQL, nil, nil); err == nil {
			if len(fallbackResult.Rows) > 0 && len(fallbackResult.Rows[0]) > 0 {
				if count, ok := fallbackResult.Rows[0][0].(int64); ok {
					totalRows = int(count)
				}
			}
		}
	}

	// Get database size
	sizeSQL := `SELECT pg_size_pretty(pg_database_size(current_database())) as size`
	sizeResult, err := dq.ExecuteQuery(ctx, sizeSQL, nil, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get database size: %w", err)
	}

	databaseSize := "Unknown"
	if len(sizeResult.Rows) > 0 && len(sizeResult.Rows[0]) > 0 {
		if size, ok := sizeResult.Rows[0][0].(string); ok {
			databaseSize = size
		}
	}

	// Get search schema specific sizes
	searchSchemaSize := "N/A"
	resourcesTableSize := "N/A"
	edgesTableSize := "N/A"

	// Get total size of all tables in search schema
	searchSchemaSizeSQL := `
		SELECT pg_size_pretty(
			COALESCE(SUM(pg_total_relation_size(schemaname||'.'||tablename)), 0)
		) as size
		FROM pg_tables
		WHERE schemaname = 'search'
	`
	if searchSchemaResult, err := dq.ExecuteQuery(ctx, searchSchemaSizeSQL, nil, nil); err == nil {
		if len(searchSchemaResult.Rows) > 0 && len(searchSchemaResult.Rows[0]) > 0 {
			if size, ok := searchSchemaResult.Rows[0][0].(string); ok {
				searchSchemaSize = size
			}
		}
	}

	// Get individual table sizes
	resourcesSizeSQL := `SELECT pg_size_pretty(pg_total_relation_size('search.resources')) as size`
	if resourcesSizeResult, err := dq.ExecuteQuery(ctx, resourcesSizeSQL, nil, nil); err == nil {
		if len(resourcesSizeResult.Rows) > 0 && len(resourcesSizeResult.Rows[0]) > 0 {
			if size, ok := resourcesSizeResult.Rows[0][0].(string); ok {
				resourcesTableSize = size
			}
		}
	}

	edgesSizeSQL := `SELECT pg_size_pretty(pg_total_relation_size('search.edges')) as size`
	if edgesSizeResult, err := dq.ExecuteQuery(ctx, edgesSizeSQL, nil, nil); err == nil {
		if len(edgesSizeResult.Rows) > 0 && len(edgesSizeResult.Rows[0]) > 0 {
			if size, ok := edgesSizeResult.Rows[0][0].(string); ok {
				edgesTableSize = size
			}
		}
	}

	// Get active connections
	connectionsSQL := `
		SELECT COUNT(*) as count
		FROM pg_stat_activity
		WHERE state = 'active'
	`
	connectionsResult, err := dq.ExecuteQuery(ctx, connectionsSQL, nil, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get connection count: %w", err)
	}

	activeConnections := 0
	if len(connectionsResult.Rows) > 0 && len(connectionsResult.Rows[0]) > 0 {
		if count, ok := connectionsResult.Rows[0][0].(int64); ok {
			activeConnections = int(count)
		}
	}

	return &DatabaseStatsResult{
		TableCount:          tableCount,
		TotalRows:          totalRows,
		DatabaseSize:       databaseSize,
		SearchSchemaSize:   searchSchemaSize,
		ResourcesTableSize: resourcesTableSize,
		EdgesTableSize:     edgesTableSize,
		ActiveConnections:  activeConnections,
	}, nil
}